tests/unit/mock.py

```diff
@@ -0,0 +1,224 @@
+from typing import Optional, Sequence, Union
+
+from errors import LxdError, RunnerError
+from github_type import RegistrationToken, RemoveToken, RunnerApplication
+from lxd_type import LxdNetwork
+from runner import LxdInstanceConfig
+
+logger = logging.getLogger(__name__)
+
+# Compressed tar file for testing.
+# Python `tarfile` module works on only files.
+# Hardcoding a sample tar file is simpler.
+TEST_BINARY = (
+    b"\x1f\x8b\x08\x00\x00\x00\x00\x00\x00\x03\xed\xd1\xb1\t\xc30\x14\x04P\xd5\x99B\x13\x04\xc9"
+    b"\xb6\xacyRx\x01[\x86\x8c\x1f\x05\x12HeHaB\xe0\xbd\xe6\x8a\x7f\xc5\xc1o\xcb\xd6\xae\xed\xde"
+    b"\xc2\x89R7\xcf\xd33s-\xe93_J\xc8\xd3X{\xa9\x96\xa1\xf7r\x1e\x87\x1ab:s\xd4\xdb\xbe\xb5\xdb"
+    b"\x1ac\xcfe=\xee\x1d\xdf\xffT\xeb\xff\xbf\xfcz\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00_{\x00"
+    b"\xc4\x07\x85\xe8\x00(\x00\x00"
+)
+
+
+class MockLxdClient:
+    """Mock the behavior of the lxd client."""
+
+    def __init__(self):
+        self.instances = MockLxdInstanceManager()
+        self.profiles = MockLxdProfileManager()
+        self.networks = MockLxdNetworkManager()
+
+
+class MockLxdInstanceManager:
+    """Mock the behavior of the lxd Instances."""
+
+    def __init__(self):
+        self.instances = {}
+
+    def create(self, config: LxdInstanceConfig, wait: bool = False) -> MockLxdInstance:
+        self.instances[config["name"]] = MockLxdInstance(config["name"])
+        return self.instances[config["name"]]
+
+    def get(self, name: str):
+        return self.instances[name]
+
+    def all(self):
+        return [i for i in self.instances.values() if not i.deleted]
+
+
+class MockLxdProfileManager:
+    """Mock the behavior of the lxd Profiles."""
+
+    def __init__(self):
+        self.profiles = set()
+
+    def create(self, name: str, config: dict[str, str], devices: dict[str, str]):
+        self.profiles.add(name)
+
+    def exists(self, name) -> bool:
+        return name in self.profiles
+
+
+class MockLxdNetworkManager:
+    """Mock the behavior of the lxd networks"""
+
+    def __init__(self):
+        pass
+
+    def get(self, name: str) -> LxdNetwork:
+        return LxdNetwork(
+            "lxdbr0", "", "bridge", {"ipv4.address": "10.1.1.1/24"}, True, ("default")
+        )
+
+
+class MockLxdInstance:
+    """Mock the behavior of a lxd Instance."""
+
+    def __init__(self, name: str):
+        self.name = name
+        self.status = "Stopped"
+        self.deleted = False
+
+        self.files = MockLxdInstanceFileManager()
+
+    def start(self, wait: bool = True, timeout: int = 60):
+        self.status = "Running"
+
+    def stop(self, wait: bool = True, timeout: int = 60):
+        self.status = "Stopped"
+        # Ephemeral virtual machine should be deleted on stop.
+        self.deleted = True
+
+    def delete(self, wait: bool = True):
+        self.deleted = True
+
+    def execute(self, cmd: Sequence[str], cwd: Optional[str] = None) -> tuple[int, str, str]:
+        return 0, "", ""
```

Comment:
Couldn't exit code be a parameter to test the behavior in case of error?

Reply:
The current method in the tests is to overwrite the method with a different stub method.